store/cachekvstore.go

```diff
@@ -0,0 +1,172 @@
+package store
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+)
+
+// If value is nil but deleted is false, it means the parent doesn't have the
+// key.  (No need to delete upon Write())
+type cValue struct {
+	value   []byte
+	deleted bool
+	dirty   bool
+}
+
+// cacheKVStore wraps an in-memory cache around an underlying KVStore.
+type cacheKVStore struct {
+	mtx    sync.Mutex
+	cache  map[string]cValue
+	parent KVStore
+}
+
+var _ CacheKVStore = (*cacheKVStore)(nil)
+
+func NewCacheKVStore(parent KVStore) *cacheKVStore {
+
+	ci := &cacheKVStore{
+		cache:  make(map[string]cValue),
+		parent: parent,
+	}
+
+	return ci
+}
+
+func (ci *cacheKVStore) Get(key []byte) (value []byte) {
+	ci.mtx.Lock()
+	defer ci.mtx.Unlock()
+	ci.assertValidKey(key)
+
+	cacheValue, ok := ci.cache[string(key)]
+	if !ok {
```

Comment:
I think I recall a familiar PR from tmlibs, but I recall commenting that let's try to also check if deleted so
```go
if !ok || cacheValue.deleted {
```